8332697: ubsan: shenandoahSimpleBitMap.inline.hpp:68:23: runtime error: signed integer overflow: -9223372036854775808 - 1 cannot be represented in type 'long int'

Henry-Lin-A opened the pull request:
Cast the result of `nth_bit(n)` to `uintptr_t` to prevent signed integer overflow error reported by `ubsan`. Unsigned overflow is not undefined behavior and is not checked by `ubsan`.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8332697](https://bugs.openjdk.org/browse/JDK-8332697): ubsan: shenandoahSimpleBitMap.inline.hpp:68:23: runtime error: signed integer overflow: -9223372036854775808 - 1 cannot be represented in type 'long int' (**Bug** - P4)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20164/head:pull/20164` \
`$ git checkout pull/20164`

Update a local copy of the PR: \
`$ git checkout pull/20164` \
`$ git pull https://git.openjdk.org/jdk.git pull/20164/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20164`

View PR using the GUI difftool: \
`$ git pr show -t 20164`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20164.diff">https://git.openjdk.org/jdk/pull/20164.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20164#issuecomment-2240184491)

Henry-Lin-A commented:
/covered

dean-long commented on src/hotspot/share/utilities/globalDefinitions.hpp:
```diff
@@ -1066,7 +1066,7 @@ const intptr_t OneBit     =  1; // only right_most bit set in a word
 // (note: #define used only so that they can be used in enum constant definitions)
 #define nth_bit(n)        (((n) >= BitsPerWord) ? 0 : (OneBit << (n)))
-#define right_n_bits(n)   (nth_bit(n) - 1)
+#define right_n_bits(n)   ((uintptr_t) nth_bit(n) - 1)
```
This changes the return type of right_n_bits, which could break existing code.  If we need an unsigned version, I think it should have a different name.

theRealAph replied:
And this is at best a partial fix: `OneBit << 63` overflows.

Henry-Lin-A commented:
Thanks for the feedback. I reverted my changes in `globalDefinitions.hpp` and added an unsigned version of `right_n_bits` in `shenandoahSimpleBitMap.hpp`. This new unsigned version replaces the usages that caused undefined overflow behavior in the `shenandoahSimpleBitMap` files.
